adoptopenjdk-api-v3-updater/src/test/kotlin/net/adoptopenjdk/api/TestResourceDouble.kt

```diff
@@ -0,0 +1,191 @@
+import net.adoptopenjdk.api.v3.dataSources.models.FeatureRelease
+import net.adoptopenjdk.api.v3.models.Architecture
+import net.adoptopenjdk.api.v3.models.Binary
+import net.adoptopenjdk.api.v3.models.HeapSize
+import net.adoptopenjdk.api.v3.models.ImageType
+import net.adoptopenjdk.api.v3.models.Installer
+import net.adoptopenjdk.api.v3.models.JvmImpl
+import net.adoptopenjdk.api.v3.models.OperatingSystem
+import net.adoptopenjdk.api.v3.models.Package
+import net.adoptopenjdk.api.v3.models.Project
+import net.adoptopenjdk.api.v3.models.Release
+import net.adoptopenjdk.api.v3.models.ReleaseType
+import net.adoptopenjdk.api.v3.models.SourcePackage
+import net.adoptopenjdk.api.v3.models.Vendor
+import net.adoptopenjdk.api.v3.models.VersionData
+import java.time.ZonedDateTime
+import java.util.Random
+
+object TestResourceDouble {
+
+    val rand: Random = Random(1)
+    val TEST_VERSIONS = listOf(8, 10, 11, 12)
+    val TEST_RESOURCES = listOf(
+        PermittedValues(
+            ReleaseType.values().asList(),
+            listOf(Vendor.adoptopenjdk),
+            listOf(Project.jdk, Project.jfr),
+            listOf(JvmImpl.hotspot),
+            ImageType.values().asList(),
+            Architecture.values().asList(),
+            OperatingSystem.values().asList(),
+            listOf(HeapSize.normal)
+        ),
+        PermittedValues(
+            ReleaseType.values().asList(),
+            listOf(Vendor.adoptopenjdk),
+            listOf(Project.jdk),
+            listOf(JvmImpl.openj9),
+            ImageType.values().asList(),
+            Architecture.values().asList(),
+            OperatingSystem.values().asList(),
+            HeapSize.values().asList()
+        ),
+        PermittedValues(
+            ReleaseType.values().asList(),
+            listOf(Vendor.openjdk),
+            listOf(Project.jdk),
+            listOf(JvmImpl.hotspot),
+            ImageType.values().asList(),
+            Architecture.values().asList(),
+            OperatingSystem.values().asList(),
+            listOf(HeapSize.normal)
+        )
+    )
+
+    fun generate(): AdoptRepos {
+        return AdoptRepos(
+            TEST_VERSIONS.associateWith { version ->
+                FeatureRelease(version, createRepos(version))
+            }
+        )
+    }
+
+    private fun createRepos(majorVersion: Int): List<AdoptRepo> {
+        return (1..2)
+            .flatMap {
+                TEST_RESOURCES.map { AdoptRepo(it.createReleases(majorVersion)) }
+            }
+            .toList()
+    }
+
+    class PermittedValues(
+        val releaseType: List<ReleaseType>,
+        val vendor: List<Vendor>,
+        val project: List<Project>,
+        val jvmImpl: List<JvmImpl>,
+        val imageType: List<ImageType>,
+        val architecture: List<Architecture>,
+        val operatingSystem: List<OperatingSystem>,
+        val heapSize: List<HeapSize>
+    ) {
+
+        fun createReleases(majorVersion: Int): List<Release> {
```

Comment:
Discussed with John - FlatMap pyramid of doom - maybe Kotlin has an operation we're missing?  @nickebbitt ?

Reply:
I'm not aware of anything Kotlin specific that would help here, I suppose we're trading off generating the test dynamically against creating it all more explicitly. 

The explicit approach is less complexity but more maintenance to extend the test data when new variations come along, whereas the dynamic approach takes more effort to understand but, in theory, should be lower maintenance.

One option would be to fall back to more standard `forEach`, might easier to grok and a simpler execution model e.g.

```
   fun createReleases(majorVersion: Int): List<Release> {
            val versions = getVersions(majorVersion)
            val releases = ArrayList<Release>()
            versions.forEach { versionData ->
                releaseType.forEach { releaseType ->
                    vendor.forEach { vendor ->
                        releases.add(Release(
                            randomString("release id"),
                            releaseType,
                            randomString("release lin"),
                            randomString("release name"),
                            randomDate(),
                            randomDate(),
                            getBinaries(),
                            2,
                            vendor,
                            versionData,
                            sourcePackage()
                        ))
                    }
                }
            }
            return releases
        }
```